Bodies structure

tjstienstra opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
This PR is part of the mechanics experimental development (#24234)

#### Brief description of what is fixed or changed
This PR implements a revised bodies structure, where `Particle` and `RigidBody` both inherit from an abstract base class `BodyBase`. This improves the similarity between the objects, so they will also be usable in the joints framework.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

tjstienstra committed:
Implement an abstract body for Rigidbody and Particle

tjstienstra committed:
Implement an abstract body for Rigidbody and Particle

tjstienstra committed:
Refactor to use formatted math in the docstrings

tjstienstra committed:
Refactor to use formatted math in the docstrings

TJStienstra commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+from abc import ABC, abstractmethod
```
Named the file `abstract_body.py`, because we are currently using separate files for all bodies and this way we can avoid circular import errors. Another option would be to put the bodies in the same `bodies.py` file, but not sure if this is allowed with respect to backward compatibility. Besides of the fact that the file will be long.

P.S. A better name is also welcome.

brocksam replied:
I think it's better to keep the files smaller as it makes navigating easier. So I support this being a separate file. I think I would prefer if there was consistency between the class name and the file name. If we're calling the ABC `_Body` then can this file be `_body.py`? Seeing as it shouldn't ever need to be used as a module by a user?

TJStienstra replied:
Not sure if `_body.py` is the best name, because we also use the `_` prefix for experimental/sandbox modules.

brocksam replied:
I agree with the point about the experimental/sandbox modules being prefixed with an underscore. I do think it's easier to navigate when classes and modules have the same name.

TJStienstra replied:
Would you consider `body_abc.py` as better?

brocksam replied:
Yes, I think I prefer `body_abc.py`. Would be nice if we followed a similar naming pattern for all ABCs. Or alternatively, we could aim to move `_Body` into `body.py` once `Body` has been deprecated and the circular dependency can be avoided another way. E.g. moving `Body` into `deprecated/body.py` or something.

TJStienstra replied:
Not sure if a separate folder for deprecated modules is ideal. It does change the import path if someone were to be using `sympy.physics.mechanics.body` (not that I would encourage that).

When looking through SymPy for abstract base classes there does not seem to be a clear approach what we should do here. Also since almost none of them inherit `ABC`. Another option for naming would be `BodyBase`, which is in naming similar to `MatrixBase`, `StateBase`, `KetBase`, `BraBase`. An option would then also be to use `body_base.py`.

brocksam replied:
> Not sure if a separate folder for deprecated modules is ideal. It does change the import path if someone were to be using sympy.physics.mechanics.body (not that I would encourage that).

Not sure that this is an issue if we're following the proper deprecation procedure. But this is now a discussion for a later date!

> When looking through SymPy for abstract base classes there does not seem to be a clear approach what we should do here. Also since almost none of them inherit `ABC`. Another option for naming would be `BodyBase`, which is in naming similar to `MatrixBase`, `StateBase`, `KetBase`, `BraBase`. An option would then also be to use `body_base.py`.

I think `BodyBase` and `body_base.py` would be my preference then if this is going to be the most consistent with elsewhere in SymPy.

TJStienstra commented on sympy/physics/mechanics/particle.py:
```diff
@@ -39,56 +41,58 @@ class Particle:
     """
+    point = _Body.masscenter
 
-    def __init__(self, name, point, mass):
```
Not sure if we would like to add a `frame` kwarg here. Disadvantage would be that the order would be different compared to `RigidBody` due to argument order backwards compatibility. Also if we decide to add the `frame` kwarg here, then we can also add it to `_Body`.

brocksam replied:
Could we add `frame` as a keyword-only argument? I.e. make the initialiser

```python
    def __init__(self, name, point=None, mass=None, *, frame=None):
```

Ideally, I'd suggest we make all arguments except `name` keyword only as that would fix the ordering inconsistency to `RigidBody` issue, but appreciate that that is likely not backwards compatible. In general I prefer keyword-only arguments in instances where there are more than one or two arguments, unless the arguments are so logical that it's not needed (e.g. `x`, `y`, `z` in a 3D vector).

TJStienstra replied:
```py
Rigidbody('B', Point('Bo'), ReferenceFrame('B'), Symbol('Bm'))
# Decide to make it a Particle
Particle('B', Point('Bo'), ReferenceFrame('B'), Symbol('Bm'))
# if key-word only
raise TypeError: Particle.__init__() takes from 2 to 4 positional arguments but 5 were given
# frame as normal key-word
raise ValueError: Frame must be an instance of ReferenceFrame.
```
Personally think that having it as a normal keyword can give quite a nice error. Yes, this will mean that they do have different order of arguments, but assuming someone uses them as keyword, then this will not pose a problem.

P.S. Will be changing the `frame = ReferenceFrame(frame)` line in the `frame.setter` as throwing a `ValueError` is more sensible.

brocksam replied:
That's poor error messaging on Python's part.

TJStienstra replied:
Just added the `frame` kwarg.

TJStienstra commented on sympy/physics/mechanics/particle.py:
```diff
@@ -39,56 +41,58 @@ class Particle:
+            self._frame = ReferenceFrame(frame)
+        self.point.set_vel(frame, 0)
+
+    def add_frame(self, frame=None):
```
Is `attach_frame` a better name?

brocksam replied:
Don't think it matters too much as this method allows both an already-instantiated frame to be attached, or the creation of a new frame. On balance I think `add_frame` better sums up the potential to both attach and create. It's also shorter. But the main argument is which verb is used elsewhere? If add is used elsewhere then good. If attach is used elsewhere then perhaps we should prefer attach for consistency.

TJStienstra replied:
Personally think `add` is indeed better, as `add` is commonly used as prefix.

TJStienstra commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -51,21 +53,24 @@ class RigidBody:
+    def __init__(self, name, masscenter=None, frame=None, mass=None,
+                 inertia=None):
+        # This import will be removed with the introduction of InertiaTuple
+        from sympy.physics.mechanics.functions import inertia as f_inertia
```
Would like to move the `inertia` function to a separate `inertia.py` file, when introducing the `InertiaTuple`. This will also remove some circular import problems.

brocksam replied:
Agree.

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+__all__ = ['_Body']
+
+
+class _Body(ABC):
```
Are you happy that this is the name we'd like to go for? I've worked on other projects where the convention has been to call this `_Body` and others where it has been `BodyABC`. Is there anywhere else in SymPy where ABCs are used and we can make sure we're consistent? I will also make sure that I'm consistent with what we decide here for the ABCs in `sympy.physics.biomechanics`.

TJStienstra replied:
The abstract method for `KanesMethod` and `LagrangeMethod` is also called `_Method`, which is in the file `method.py`

brocksam replied:
If we're going with `BodyBase` then it would make sense for `_Method` to become `MethodBase`. I assume this isn't problematic because the leading underscore indicates that it's private anyway. But something for a different PR.

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+class _Body(ABC):
+    """Abstract class for body type objects."""
+    def __init__(self, name, masscenter=None, mass=None):
+        if not isinstance(name, str):
+            raise TypeError('Supply a valid name.')
```
Do we want to do any further error checking here? For example ensuring that the name has nonzero length? Or does this break backwards compatibility?

brocksam replied:
Okay, we can leave as is for now. But perhaps when we implement `System` it'll be a good time to do error checking on name clashes.

TJStienstra replied:
Yeah, in `System` there will definitely be checking of all these kind of things.

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+        self.potential_energy = 0
+        self.points = []
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return self.__str__()
```
These should be improved, but again I assume they are currently like this to meet backwards compatibility?

TJStienstra replied:
This is the backward compatible variant, but in case of the `__repr__` I'm personally fine with making it a bit better. Do note that it is not possible to exactly create a string, which would replicate the body, as frame and masscenter are both part of graphs.

brocksam replied:
I think even if we can't make `__repr__` produce a string to directly recreate an instance, we can still make it much more informative than it currently is.

oscarbenjamin replied:
If only `__repr__` is provided then `str(obj)` will use it so there shouldn't be a need to have one method call the other.

TJStienstra replied:
Implemented an improved `__repr__` method for all body objects.

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+    def __repr__(self):
+        return self.__str__()
+
+    @property
+    def name(self):
+        return self._name
```
Docstring?

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,136 @@
+        """The ReferenceFrame fixed to the body."""
+        pass
+
+    @frame.setter
+    @abstractmethod
+    def frame(self, frame):
+        pass
```
Docstring?

TJStienstra replied:
Should the `setter` also have a docstring, if the `property` already has one? This is most of the time not done in SymPy.

brocksam replied:
No. This was my mistake. I didn't realise that this was the setter.

brocksam commented on sympy/physics/mechanics/body.py:
```diff
@@ -99,7 +99,6 @@ class Body(RigidBody, Particle):  # type: ignore
     def __init__(self, name, masscenter=None, mass=None, frame=None,
                  central_inertia=None):
```
Do we want to add any sort of deprecation note on instantiation?

TJStienstra replied:
`Body` is not deprecated yet in this PR, that will only be done after the joints framework has been made fully compatible with `Rigidbody` and `Particle`. Also would like to add the deprecation in a separate PR, because quite some documentation will have to be changed as well.

brocksam commented on sympy/physics/mechanics/particle.py:
```diff
@@ -39,56 +41,58 @@ class Particle:
     >>> # Or you could change these later
     >>> pa.mass = m
     >>> pa.point = po
+    >>> # If preferred you can also attach a frame to a Particle
```
```suggestion
    >>> # It is possible to attach a frame to a Particle if desired
```

brocksam commented on sympy/physics/mechanics/particle.py:
```diff
@@ -39,56 +41,58 @@ class Particle:
+
+    def add_frame(self, frame=None):
+        """Method to attach a frame to the particle."""
+        if frame is None:
+            if self._frame is not None:
+                return
+            frame = ReferenceFrame(f'{self.name}_frame')
+        self.frame = frame
```
Would it be nice for this method to return the newly-added frame?

TJStienstra replied:
Yeah, could do. Note that this method does not always create a new frame (if no frame is supplied and there is already a frame).

brocksam commented on sympy/physics/mechanics/particle.py:
```diff
@@ -157,20 +162,22 @@ def angular_momentum(self, point, frame):
         """
 
-        return self.point.pos_from(point) ^ (self.mass * self.point.vel(frame))
+        return cross(self.point.pos_from(point),
+                     (self.mass * self.point.vel(frame)))
```
Potential to remove unnecessary parentheses. I think it makes the indentation slightly easier to read.
```suggestion
        return cross(self.point.pos_from(point),
                     self.mass * self.point.vel(frame))
```

brocksam commented on sympy/physics/mechanics/particle.py:
```diff
@@ -116,19 +120,20 @@ def linear_momentum(self, frame):
-        the inertial frame, N.
+        where $r$ is the position vector from point ``O`` to the particle ``P``,
+        $m$ is the mass of the particle, and $v$ is the velocity of the particle
+        in the inertial frame, ``N``.
```
```suggestion
        in the inertial frame ``N``.
```

brocksam commented on sympy/physics/mechanics/particle.py:
```diff
@@ -157,20 +162,22 @@ def angular_momentum(self, point, frame):
         ===========
 
-        The kinetic energy, T, of a particle, P, is given by
+        The kinetic energy, $T$, of a particle, ``P``, is given by
```
Removing commas for consistency with other docstrings.
```suggestion
        The kinetic energy $T$, of a particle ``P``, is given by
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -132,18 +117,19 @@ def central_inertia(self, I):
 
-        where M is the mass of the rigid body and v* is the velocity of
-        the mass center of B in the frame, N.
+        where $m$ is the mass of the rigid body and $v$ is the velocity of
+        the mass center of ``B`` in the frame, ``N``.
```
```suggestion
        where $m$ is the mass of the rigid body, and $v$ is the velocity of
        the mass center of ``B`` in the frame ``N``.
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -173,21 +159,22 @@ def linear_momentum(self, frame):
-        of body B in the frame, N, r is the position vector from point O to the
-        mass center of B, and v is the velocity of the mass center in the
-        frame, N.
+        where $I$ and $m$ are the central inertia dyadic and mass of rigid body
+        ``B``, $\omega$ is the angular velocity of body ``B`` in the frame,
+        ``N``, $r$ is the position vector from point ``O`` to the mass center of
+        ``B``, and $v$ is the velocity of the mass center in the frame, ``N``.
```
```suggestion
        where $I$ and $m$ are the central inertia dyadic and mass of rigid body
        ``B``, $\omega$ is the angular velocity of body ``B`` in the frame
        ``N``, $r$ is the position vector from point ``O`` to the mass center of
        ``B``, and $v$ is the velocity of the mass center in the frame ``N``.
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -225,26 +212,28 @@ def angular_momentum(self, point, frame):
         ===========
 
-        The kinetic energy, T, of a rigid body, B, is given by
+        The kinetic energy, $T$, of a rigid body, ``B``, is given by
```
```suggestion
        The kinetic energy $T$, of a rigid body ``B``, is given by
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -225,26 +212,28 @@ def angular_momentum(self, point, frame):
-        where I and m are the central inertia dyadic and mass of rigid body B,
-        respectively, omega is the body's angular velocity and v is the
-        velocity of the body's mass center in the supplied ReferenceFrame.
+        where $I$ and $m$ are the central inertia dyadic and mass of rigid body
+        ``B``, respectively, omega is the body's angular velocity and $v$ is the
+        velocity of the body's mass center in the supplied ``ReferenceFrame``.
```
```suggestion
        where $I$ and $m$ are the central inertia dyadic and mass of rigid body
        ``B`` respectively, $\omega$ is the body's angular velocity, and $v$ is the
        velocity of the body's mass center in the supplied ``ReferenceFrame``.
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -266,65 +255,13 @@ def kinetic_energy(self, frame):
-        translational_KE = (self.mass * (self.masscenter.vel(frame) &
-            self.masscenter.vel(frame)) / sympify(2))
-
+        rotational_KE = dot(self.frame.ang_vel_in(frame), dot(
+            self.central_inertia, self.frame.ang_vel_in(frame)) / sympify(2))
```
I think this is easier to read.
```suggestion
        rotational_KE = dot(
            self.frame.ang_vel_in(frame),
            dot(self.central_inertia, self.frame.ang_vel_in(frame)) / sympify(2)
        )
```

brocksam commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -266,65 +255,13 @@ def kinetic_energy(self, frame):
-
+        rotational_KE = dot(self.frame.ang_vel_in(frame), dot(
+            self.central_inertia, self.frame.ang_vel_in(frame)) / sympify(2))
+        translational_KE = (self.mass * dot(
+            self.masscenter.vel(frame), self.masscenter.vel(frame)) / sympify(2)
+                            )
```
I think this is easier to read. Would require import of `S.Half`.
```suggestion
        translational_KE = S.Half * self.mass * dot(
            self.masscenter.vel(frame), self.masscenter.vel(frame)
        )
                            
```

brocksam commented on sympy/physics/mechanics/tests/test_particle.py:
```diff
@@ -1,14 +1,23 @@
 
 
 def test_particle():
+    # Test default
+    p = Particle('P')
+    assert p.name == 'P'
+    assert p.mass == Symbol('P_mass')
+    assert p.masscenter.name == 'P_masscenter'
+    raises(AttributeError, lambda: p.frame)
```
Could this be a separate test? Called `test_particle_default` or something?

brocksam commented on sympy/physics/mechanics/tests/test_particle.py:
```diff
@@ -44,6 +53,28 @@ def test_particle():
 
 
+def test_particle_frame():
+    # Test default
+    p = Particle('P')
+    raises(AttributeError, lambda: p.frame)
```
This is tested in `test_particle`.

brocksam replied:
Can be removed, especially if refactoring the above as per suggestion.

brocksam commented on sympy/physics/mechanics/tests/test_rigidbody.py:
```diff
@@ -1,13 +1,22 @@
 
 
 def test_rigidbody():
+    # Test default
+    b = RigidBody('B')
+    I = inertia(b.frame, *symbols('B_ixx B_iyy B_izz B_ixy B_iyz B_izx'))
+    assert b.name == 'B'
+    assert b.mass == Symbol('B_mass')
+    assert b.masscenter.name == 'B_masscenter'
+    assert b.inertia == (I, b.masscenter)
+    assert b.central_inertia == I
+    assert b.frame.name == 'B_frame'
```
Can we refactor to a separate test `test_rigidbody_default`?

brocksam reviewed (commented):
General comment about `abstract_body.py`: these is inconsistency between the properties over which support setters and which don't. If we want to allow users to change the value of a property after instantiation then a setter is okay. If we want properties to be immutable then we should either protect against this in the setter or just have the private attribute associated with the property set in the initialiser.

tjstienstra committed:
Add frame keyword to Particle initializer

tjstienstra committed:
Add frame keyword to Particle initializer

TJStienstra commented:
> these is inconsistency between the properties over which support setters and which don't.

@brocksam which properties do you mean? Can you clarify your comment a bit further?

tjstienstra committed:
Refactor and return frame in Particle.add_frame

tjstienstra committed:
Refactor and return frame in Particle.add_frame

TJStienstra commented:
Just notice that the documentation is crashing on the problem that `potential_energy` has been moved to the abstract class `_Body`. Should I put in something like this to fix it?
```py
@property
def potential_energy(self):
    """docstring"""
    return _Body.potential_energy.fget(self)

@potential_energy.setter
def potential_energy(self, potential_energy):
    _Body.potential_energy.fset(self, potential_energy)
```
Would like a cleaner option.

brocksam commented on sympy/physics/mechanics/body.py:
```diff
@@ -129,13 +129,14 @@ def __init__(self, name, masscenter=None, mass=None, frame=None,
 
         # If user passes masscenter and mass then a particle is created
         # otherwise a rigidbody. As a result a body may or may not have inertia.
+        # Note: _Body.__init__ is used to prevent problems with super() calls in
+        # Particle and Rigidbody.
```
```suggestion
        # Note: _Body.__init__ is used to prevent problems with super() calls in
        # Particle and RigidBody.
```

brocksam commented:
> > these is inconsistency between the properties over which support setters and which don't.
> 
> @brocksam which properties do you mean? Can you clarify your comment a bit further?

Ignore this. Having reread `_Body` just now I can't figure out what I was on about here now. I agree that it's sensible for `name` and consistent for `mass`, `masscenter` and `frame`.

tjstienstra committed:
Add inherited properties to online documentation of Particle and RigidBody

tjstienstra committed:
Add inherited properties to online documentation of Particle and RigidBody

tjstienstra committed:
Improve repr of the bodies

tjstienstra committed:
Improve repr of the bodies

tjstienstra committed:
Improve explanation of a comment

tjstienstra committed:
Improve explanation of a comment

brocksam commented on sympy/physics/mechanics/abstract_body.py:
```diff
@@ -0,0 +1,123 @@
+        # Note: If frame=None, no auto-generated frame is created, because a
+        # Particle does not need to have a frame by default.
+        if not isinstance(name, str):
+            raise TypeError('Supply a valid name.')
```
This isn't the most useful error message. Can we make it clearer?
```suggestion
            raise TypeError(f'Supplied name {name} is invalid, must be a {str}.')
```

TJStienstra replied:
Similar error messaging is used for `Joint`, `ReferenceFrame` and `Point`

brocksam replied:
Okay, ignore for now. Could be improved in the future.

tjstienstra committed:
Rename abstract body class to BodyBase

tjstienstra committed:
Rename abstract body class to BodyBase

TJStienstra commented:
Just rebased the branch on `master`, so the `mypy` tests should now also work.

tjstienstra committed:
Merge branch 'BodiesStructure' of github.com:TJStienstra/sympy into BodiesStructure

tjstienstra committed:
Update docstring equations to ascii notation as discussed in #24451

TJStienstra commented:
Sorry, this PR has been sitting still for some time. Just rebased it to be up to date with the master and implemented the suggestion in #24451 (see [this comment](https://github.com/sympy/sympy/issues/24451#issuecomment-1376984182)).

tjstienstra committed:
Fix error when initializing a particle while supplying a frame argument

brocksam commented:
I'm not sure this rebase has worked as intended. It's not going to be feasible to review like this given that 312 files have changed and 10s of 1000s of lines have been added and deleted.

asmeurer commented:
As I noted on another PR, I suspect your master branch itself has gotten corrupt. So you should 

1. Use something like `git checkout master; git reset --hard origin/master` to reset your master to the origin/master (assuming you cloned from sympy/sympy). Make sure you don't have any uncommitted changes before doing this as this will irrevocably delete them if you do. 
2. Once you've confirmed that `git show master` is the same SHA1 as shown at https://github.com/sympy/sympy/commits/master, use `git rebase --onto` or `git rebase --interactive` to fix this branch (and your other ones).

In the future, be very careful using commands like rebase, or this sort of thing can happen. If you just use `merge`, it won't happen. You should never rebase or do any sort of commiting on `master` itself. Only pull the upstream changes to master. On your other branches, avoid rebasing commits from master (e.g., don't use things like `git rebase HEAD~10` which might pickup commits from master). Personally I never use rebase and only merge (and then, only merge when needed to get upstream changes). If you do that, you'll never have this problem.

tjstienstra committed:
Merge branch 'physics-mechanics-devel' into BodiesStructure

# Conflicts:
#	sympy/physics/mechanics/particle.py
#	sympy/physics/mechanics/rigidbody.py

TJStienstra commented on sympy/physics/mechanics/rigidbody.py:
```diff
@@ -140,10 +129,10 @@ def linear_momentum(self, frame):
         given by:
 
-        L = M * v*
+        ``L = m * v``
```
Also changed it to lowercase, as was a bit later discussed in #24451

tjstienstra committed:
Swap the dyadic explanation between physic.vector and physics.mechanics

tjstienstra committed:
Fix reference target not found

brocksam commented on sympy/physics/mechanics/body_base.py:
```diff
@@ -0,0 +1,123 @@
+        self.potential_energy = 0
+        self.points = []
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return (f'{self.__class__.__name__}({repr(self.name)}, masscenter='
+                f'{repr(self.masscenter)}, frame={repr(self.frame)}, mass='
+                f'{repr(self.mass)})')
```
Non-essential, but I personally like seeing `__str__` and `__repr__` implementations at the very end of a class definition. This may not be the case in SymPy though, so if not please ignore.

TJStienstra replied:
The location of magic methods in SymPy seems to be inconsistent. I've always put them at the start after the `__init__`, same seems to happen in for example `Point`, but I can also find some examples putting them at the end.

brocksam commented on sympy/physics/mechanics/tests/test_particle.py:
```diff
@@ -1,15 +1,30 @@
+    assert p.masscenter.name == 'P_masscenter'
+    assert p.potential_energy == 0
+    assert p.__str__() == 'P'
+    assert p.__repr__() == ("Particle('P', masscenter=P_masscenter, frame=None,"
```
Is this correct/are we happy with this? The repr should ideally return the code that, when pasted into the console, reinstantiates the same object. `P_masscenter` is just the `Particle`'s name as not the repr for `p`.

TJStienstra replied:
`P_masscenter` would be something that should then be fixed in the `Point` class. The difficulty here would also lie in the fact that a `Point` has an internal dictionaries (as it is a node in multiple graphs).
